File: packages/modules/haptics/ios/HapticsModule.swift
import StingRuntime
import UIKit

public final class HapticsModule: StingNativeModule {
    public let name = "Haptics"
    public let version = "0.1.0"

    public init() {}

    public func callSync(method: String, arguments: [Any]) throws -> Any? {
        guard method == "impact" else {
            throw StingNativeModuleError(
                code: "E_METHOD_NOT_FOUND",
                message: "Haptics does not implement \(method)"
            )
        }

        let rawStyle = arguments.first as? String ?? "medium"
        let style: UIImpactFeedbackGenerator.FeedbackStyle
        switch rawStyle {
        case "light": style = .light
        case "heavy": style = .heavy
        case "soft": style = .soft
        case "rigid": style = .rigid
        default: style = .medium
        }

        let generator = UIImpactFeedbackGenerator(style: style)
        generator.prepare()
        generator.impactOccurred()
        return nil
    }
}

File: packages/modules/haptics/android/src/main/java/run/stingjs/modules/haptics/HapticsModule.kt
package run.stingjs.modules.haptics

import android.content.Context
import android.os.Build
import android.os.VibrationEffect
import android.os.Vibrator
import android.os.VibratorManager
import run.stingjs.runtime.StingNativeModule
import run.stingjs.runtime.StingNativeModuleError

class HapticsModule(context: Context) : StingNativeModule {
    override val name = "Haptics"
    override val version = "0.1.0"

    private val appContext = context.applicationContext

    override fun callSync(method: String, arguments: List<Any?>): Any? {
        if (method != "impact") {
            throw StingNativeModuleError(
                code = "E_METHOD_NOT_FOUND",
                message = "Haptics does not implement $method",
            )
        }

        val style = arguments.firstOrNull() as? String ?: "medium"
        val (durationMs, amplitude) = when (style) {
            "light" -> 12L to 70
            "heavy" -> 28L to 220
            "soft" -> 18L to 90
            "rigid" -> 16L to 180
            else -> 20L to 140
        }

        vibrate(durationMs, amplitude)
        return null
    }

    @Suppress("DEPRECATION")
    private fun vibrate(durationMs: Long, amplitude: Int) {
        val vibrator = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            appContext.getSystemService(VibratorManager::class.java).defaultVibrator
        } else {
            appContext.getSystemService(Context.VIBRATOR_SERVICE) as Vibrator
        }

        if (!vibrator.hasVibrator()) return

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            vibrator.vibrate(VibrationEffect.createOneShot(durationMs, amplitude))
        } else {
            vibrator.vibrate(durationMs)
        }
    }
}
